File: plugin/testData/shared/basics/ComplexGenerics.shared.kt
package com.test

import com.lightningkite.khrysalis.*

data class NameTag<T: Hashable>(val item: T, val name: String)

fun <T: MyInterface> NameTag<T>.printInterface(){
    println("Hello!  I am $name, I stand for ${item.x}.")
}

interface MyInterface {
    val x: String get() = ""
    fun y(string: String): String {
        return x + string
    }
}

data class ImplOverX(val k: Int = 0): MyInterface {
    override val x: String
        get() = "Hello!"
}

data class ImplOverY(val k: Int = 0): MyInterface {
    override fun y(string: String): String {
        return "$x!"
    }
}

data class ImplBoth(val k: Int = 0): MyInterface {
    override val x: String
        get() = "Hello!"
    override fun y(string: String): String {
        return "$x!"
    }
}

fun main(){
    NameTag(ImplBoth(), "ImplBoth").printInterface()
    NameTag(ImplOverX(), "ImplOverX").printInterface()
    NameTag(ImplOverY(), "ImplOverY").printInterface()
}

File: plugin/testDataOutputTs/shared/basics/ComplexGenerics.shared.ts
//Package: com.test
//Converted using Khrysalis2


class NameTag<T extends Hashable> {
    item: T;
    name: string;
    
    public constructor(item: T, name: string) {
        this.item = item;
        this.name = name;
    }
    
}

 

export function printInterface<T extends MyInterface>(): void {
    console.log(`Hello!  I am ${this.name}, I stand for ${this.item.x}.`)
}

 

interface MyInterface {
    implementsInterfaceMyInterface: boolean;
    readonly x : string;
    y(aString: string): string ;
}

 

class ImplOverX {
    k: number;
    
    public constructor(k: number = (0)) {
        this.k = k;
    }
    
    public x : string;
}

 

class ImplOverY {
    k: number;
    
    public constructor(k: number = (0)) {
        this.k = k;
    }
    
    y(aString: string): string {
        return `${this.x}!`
    }
}

 

class ImplBoth {
    k: number;
    
    public constructor(k: number = (0)) {
        this.k = k;
    }
    
    public x : string;
    y(aString: string): string {
        return `${this.x}!`
    }
}

 

export function main(): void {
    (new NameTag(new ImplBoth(), "ImplBoth")).printInterface();
    (new NameTag(new ImplOverX(), "ImplOverX")).printInterface();
    (new NameTag(new ImplOverY(), "ImplOverY")).printInterface()
}

 
